controllers/apps/transformer_cluster_load_resources.go

```diff
@@ -41,136 +41,147 @@ func (t *clusterLoadRefResourcesTransformer) Transform(ctx graph.TransformContex
+			return fmt.Errorf("component %s not defined in topology %s", comp.Name, clusterTopology.Name)
+		}
 	}
 
+	cluster.Spec.Topology = clusterTopology.Name
+
 	return nil
 }
 
-func (t *clusterLoadRefResourcesTransformer) allCompDefRefs(cluster *appsv1alpha1.Cluster) []string {
-	refs := make([]string, 0)
-	for _, compSpec := range cluster.Spec.ComponentSpecs {
-		if len(compSpec.ComponentDef) == 0 {
-			continue
-		}
-		refs = append(refs, compSpec.ComponentDef)
-	}
-	for _, shardingSpec := range cluster.Spec.ShardingSpecs {
-		if len(shardingSpec.Template.ComponentDef) == 0 {
-			continue
-		}
-		for i := 0; i < int(shardingSpec.Shards); i++ {
-			refs = append(refs, shardingSpec.Template.ComponentDef)
-		}
-	}
-	return refs
+func withClusterTopology(cluster *appsv1alpha1.Cluster) bool {
+	return len(cluster.Spec.ClusterDefRef) > 0 && clusterCompCnt(cluster) == clusterNewCompCnt(cluster)
 }
 
-func (t *clusterLoadRefResourcesTransformer) allCompSpecCnt(cluster *appsv1alpha1.Cluster) int {
-	compSpecCount := len(cluster.Spec.ComponentSpecs)
-	for _, shardingSpec := range cluster.Spec.ShardingSpecs {
-		compSpecCount += int(shardingSpec.Shards)
-	}
-	return compSpecCount
+func withClusterUserDefined(cluster *appsv1alpha1.Cluster) bool {
+	return len(cluster.Spec.ClusterDefRef) == 0 && len(cluster.Spec.Topology) == 0 && clusterCompCnt(cluster) == clusterNewCompCnt(cluster)
 }
 
-func (t *clusterLoadRefResourcesTransformer) loadAndCheckComponentDefinitions(
-	ctx *clusterTransformContext, cluster *appsv1alpha1.Cluster) error {
+func withClusterLegacyDefinition(cluster *appsv1alpha1.Cluster) bool {
+	return len(cluster.Spec.ClusterDefRef) > 0 && len(cluster.Spec.Topology) == 0 && clusterCompCnt(cluster) == clusterLegacyCompCnt(cluster)
+}
 
-	if ctx.ComponentDefs == nil {
-		ctx.ComponentDefs = make(map[string]*appsv1alpha1.ComponentDefinition)
-	}
+func withClusterSimplifiedAPI(cluster *appsv1alpha1.Cluster) bool {
+	return apiconversion.HasSimplifiedClusterAPI(cluster)
+}
 
-	loadAndCheck := func(compDefName string) error {
-		if len(compDefName) == 0 {
-			return nil
-		}
-		if _, ok := ctx.ComponentDefs[compDefName]; ok {
-			return nil
-		}
-		compDef := &appsv1alpha1.ComponentDefinition{}
-		if err := ctx.Client.Get(ctx.Context, types.NamespacedName{Name: compDefName}, compDef); err != nil {
-			return err
-		}
-		if compDef.Status.Phase != appsv1alpha1.AvailablePhase {
-			return fmt.Errorf("the component definition referenced is unavailable: %s", compDefName)
-		}
-		ctx.ComponentDefs[compDefName] = compDef
-		return nil
+func clusterCompCnt(cluster *appsv1alpha1.Cluster) int {
+	return clusterCompCntWithFunc(cluster, func(spec appsv1alpha1.ClusterComponentSpec) bool { return true })
+}
+
+func clusterNewCompCnt(cluster *appsv1alpha1.Cluster) int {
+	isNewComp := func(spec appsv1alpha1.ClusterComponentSpec) bool {
+		return len(spec.ComponentDefRef) == 0 // spec.componentDef is optional and can be inferred by operator.
```

Comment:
Why isn't it checking len(spec.ComponentDef) > 0 here? In the API documentation, ComponentDef has a higher priority than ComponentDefRef, so in the case of newComp, ComponentDef should not be empty. Assume a scenario where ClusterDefRef, ComponentDef, and ComponentDefRef are all non-empty; in this scenario, none of withClusterTopology, withClusterUserDefined, or withClusterLegacyDefinition would match.

Reply:
> Why isn't it checking len(spec.ComponentDef) > 0 here? In the API documentation, ComponentDef has a higher priority than ComponentDefRef, so in the case of newComp, ComponentDef should not be empty. Assume a scenario where ClusterDefRef, ComponentDef, and ComponentDefRef are all non-empty; in this scenario, none of withClusterTopology, withClusterUserDefined, or withClusterLegacyDefinition would match.

OK, I've roughly understood. In the new API, ComponentDef may be empty and can be inferred from the topology. However, we should still handle the aforementioned case here, which is the scenario where ClusterDefRef, ComponentDef, and ComponentDefRef are all simultaneously non-empty.

Reply:
Remove clusterNewCompCnt and use clusterLegacyCompCnt to do the check.